sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/exception/CosmosExceptionUtils.java

```diff
@@ -37,9 +38,40 @@ public static <T> Mono<T> exceptionHandler(String message, Throwable throwable,
         }
         //  Unwrap the exception in case if it is a reactive exception
         final Throwable unwrappedThrowable = Exceptions.unwrap(throwable);
-        if (unwrappedThrowable instanceof CosmosException) {
-            CosmosException cosmosException = (CosmosException) unwrappedThrowable;
+        if (unwrappedThrowable instanceof CosmosException cosmosException) {
             CosmosUtils.fillAndProcessCosmosExceptionDiagnostics(responseDiagnosticsProcessor, cosmosException);
+
+            switch (((CosmosException) unwrappedThrowable).getStatusCode()) {
+                case HttpConstants.StatusCodes.BADREQUEST -> throw new CosmosBadRequestException(message, unwrappedThrowable);
+                case HttpConstants.StatusCodes.CONFLICT -> throw new CosmosConflictException(message, unwrappedThrowable);
+                case HttpConstants.StatusCodes.FORBIDDEN -> throw new CosmosForbiddenException(message, unwrappedThrowable);
+                case HttpConstants.StatusCodes.GONE-> {
+                    switch(((CosmosException) unwrappedThrowable).getSubStatusCode()) {
+                        case HttpConstants.SubStatusCodes.NAME_CACHE_IS_STALE -> throw new CosmosInvalidPartitionException(message, unwrappedThrowable);
+                        case HttpConstants.SubStatusCodes.COMPLETING_PARTITION_MIGRATION -> throw new CosmosPartitionIsMigratingException(message, unwrappedThrowable);
+                        case HttpConstants.SubStatusCodes.PARTITION_KEY_RANGE_GONE -> throw new CosmosPartitionKeyRangeGoneException(message, unwrappedThrowable);
+                        case HttpConstants.SubStatusCodes.COMPLETING_SPLIT_OR_MERGE -> throw new CosmosPartitionKeyRangeIsSplittingException(message, unwrappedThrowable);
+                        default -> throw new CosmosGoneException(message, unwrappedThrowable);
+                    }
+                }
+                case HttpConstants.StatusCodes.INTERNAL_SERVER_ERROR -> throw new CosmosInternalServerErrorException(message, unwrappedThrowable);
+                case HttpConstants.StatusCodes.METHOD_NOT_ALLOWED -> throw new CosmosMethodNotAllowedException(message, unwrappedThrowable);
+                case HttpConstants.StatusCodes.NOTFOUND -> throw new CosmosNotFoundException(message, unwrappedThrowable);
```

Comment:
```suggestion
                case HttpConstants.StatusCodes.NOTFOUND -> throw new CosmosNotFoundException(message, cosmosException);
```
```suggestion
                case HttpConstants.StatusCodes.NOTFOUND -> throw new CosmosNotFoundException(message, unwrappedThrowable);
```

Reply:
I think you probably also need to special case 404/1002 (ReadSessionNotAvailable)? - I think this can leak form public API (and would not necessarily be mapped to a 503 or similar), because a session otken could simply be incorrect in app. And Spring dev could easily confusde/miss the substatuscode which is very significant semantic difference between real not found and ReadSessionNotAvailable.